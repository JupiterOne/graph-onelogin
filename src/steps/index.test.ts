import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../config';
import { fetchUsers } from './users';
import { fetchGroups } from './groups';
import { fetchRoles } from './roles';
import { fetchApplications } from './applications';
import { fetchUserApps } from './userapps';
import { fetchUserDevices } from './userdevices';
import { fetchAccountDetails } from './account';

import { integrationConfig } from '../../test/config';
import { setupOneloginRecording } from '../../test/recording';

let recording: Recording;

afterEach(async () => {
  await recording.stop();
});

test('should collect data', async () => {
  recording = setupOneloginRecording({
    directory: __dirname,
    name: 'steps',
  });

  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchGroups(context);
  await fetchRoles(context);
  await fetchUsers(context);
  await fetchApplications(context);
  await fetchUserApps(context);
  await fetchUserDevices(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'onelogin_account' },
        name: { type: 'string' },
        displayName: { type: 'string' },
        webLink: { type: 'string', format: 'url' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'displayName', 'webLink'],
    },
  });

  const users = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('User'),
  );
  expect(users.length).toBeGreaterThan(0);
  expect(users).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'onelogin_user' },
        name: { type: 'string' },
        displayName: { type: 'string' },
        email: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'displayName', 'email'],
    },
  });

  const appClients = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Application'),
  );
  expect(appClients.length).toBeGreaterThan(0);
  expect(appClients).toMatchGraphObjectSchema({
    _class: ['Application'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'onelogin_application' },
        name: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });
});
