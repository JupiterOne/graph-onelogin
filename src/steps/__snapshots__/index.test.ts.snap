// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "onelogin_account_local-integration-instance",
      "_type": "onelogin_account",
      "displayName": "Local Integration",
      "name": "Local Integration",
      "webLink": "https://jupiterone.onelogin.com",
    },
    Object {
      "_class": Array [
        "Service",
        "Control",
      ],
      "_key": "onelogin:sso:Local Integration",
      "_type": "onelogin_service",
      "category": "security",
      "controlDomain": "identity-access",
      "displayName": "Onelogin SSO",
      "function": "sso",
      "name": "SSO",
    },
    Object {
      "_class": Array [
        "Service",
        "Control",
      ],
      "_key": "onelogin:mfa:Local Integration",
      "_type": "onelogin_service",
      "category": "security",
      "controlDomain": "identity-access",
      "displayName": "Onelogin MFA",
      "function": "mfa",
      "name": "MFA",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "onelogin_group_477661",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": 477661,
            "name": "Getting the Band Together",
            "reference": null,
          },
        },
      ],
      "_type": "onelogin_group",
      "displayName": "Getting the Band Together",
      "id": "477661",
      "name": "Getting the Band Together",
      "reference": null,
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_428308",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": 428308,
            "name": "Default",
          },
        },
      ],
      "_type": "onelogin_role",
      "displayName": "Default",
      "id": "428308",
      "name": "Default",
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_439868",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": 439868,
            "name": "Rock Star",
          },
        },
      ],
      "_type": "onelogin_role",
      "displayName": "Rock Star",
      "id": "439868",
      "name": "Rock Star",
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_439869",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": 439869,
            "name": "Ninja",
          },
        },
      ],
      "_type": "onelogin_role",
      "displayName": "Ninja",
      "id": "439869",
      "name": "Ninja",
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_439870",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": 439870,
            "name": "Cryptozoologist",
          },
        },
      ],
      "_type": "onelogin_role",
      "displayName": "Cryptozoologist",
      "id": "439870",
      "name": "Cryptozoologist",
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_459456",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": 459456,
            "name": "AWS_BINDING",
          },
        },
      ],
      "_type": "onelogin_role",
      "displayName": "AWS_BINDING",
      "id": "459456",
      "name": "AWS_BINDING",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "onelogin_user_132422371",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "activated_at": "2021-04-28T21:10:57.207Z",
            "comment": "",
            "company": "",
            "created_at": "2021-04-28T21:10:57.210Z",
            "custom_attributes": Object {
              "customUserField": "custom-user-value",
            },
            "department": "",
            "directory_id": null,
            "distinguished_name": null,
            "email": "exampleperson@jupiterone.com",
            "external_id": null,
            "firstname": "Example",
            "group_id": null,
            "id": 132422371,
            "invalid_login_attempts": 0,
            "invitation_sent_at": "2021-04-28T21:10:57.187Z",
            "last_login": "2021-08-05T18:36:08.876Z",
            "lastname": "Person",
            "locale_code": null,
            "locked_until": null,
            "manager_ad_id": null,
            "manager_user_id": null,
            "member_of": null,
            "openid_name": "example.person",
            "password_changed_at": "2021-08-02T22:37:29.935Z",
            "phone": "",
            "preferred_locale_code": null,
            "role_id": Array [
              439868,
            ],
            "samaccountname": null,
            "state": 1,
            "status": 1,
            "title": "",
            "trusted_idp_id": null,
            "updated_at": "2021-08-05T18:36:08.984Z",
            "username": "",
            "userprincipalname": null,
          },
        },
      ],
      "_type": "onelogin_user",
      "activatedAt": 1619644257207,
      "comment": "",
      "company": "",
      "createdAt": 1619644257210,
      "customAttributes.customUserField": "custom-user-value",
      "department": "",
      "directoryId": null,
      "displayName": "Example Person",
      "distinguishedName": null,
      "email": "exampleperson@jupiterone.com",
      "externalId": null,
      "firstname": "Example",
      "groupId": 0,
      "id": "132422371",
      "invalidLoginAttempts": 0,
      "invitationSentAt": 1619644257187,
      "lastLogin": 1628188568876,
      "lastname": "Person",
      "localeCode": null,
      "lockedUntil": null,
      "managerAdId": null,
      "managerUserId": null,
      "memberOf": null,
      "name": "Example Person",
      "openidName": "example.person",
      "passwordChangedAt": 1627943849935,
      "phone": "",
      "preferredLocaleCode": null,
      "roles": "Rock Star;",
      "samaccountname": null,
      "state": 1,
      "status": "1",
      "title": "",
      "trustedIdpId": null,
      "updatedAt": 1628188568984,
      "username": "",
      "userprincipalname": null,
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "onelogin_user_133771160",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "activated_at": "2021-05-11T13:58:57.421Z",
            "comment": "",
            "company": "",
            "created_at": "2021-05-11T13:58:57.430Z",
            "custom_attributes": Object {
              "customUserField": "",
            },
            "department": "",
            "directory_id": null,
            "distinguished_name": null,
            "email": "exampleperson@jupiterone.com",
            "external_id": null,
            "firstname": "Example",
            "group_id": 477661,
            "id": 133771160,
            "invalid_login_attempts": 0,
            "invitation_sent_at": null,
            "last_login": null,
            "lastname": "Person",
            "locale_code": null,
            "locked_until": null,
            "manager_ad_id": null,
            "manager_user_id": null,
            "member_of": null,
            "openid_name": "example.person",
            "password_changed_at": null,
            "phone": "",
            "preferred_locale_code": null,
            "role_id": Array [
              439868,
              439870,
            ],
            "samaccountname": null,
            "state": 1,
            "status": 1,
            "title": "",
            "trusted_idp_id": null,
            "updated_at": "2021-06-16T18:13:37.199Z",
            "username": "",
            "userprincipalname": null,
          },
        },
      ],
      "_type": "onelogin_user",
      "activatedAt": 1620741537421,
      "comment": "",
      "company": "",
      "createdAt": 1620741537430,
      "customAttributes.customUserField": "",
      "department": "",
      "directoryId": null,
      "displayName": "Example Person",
      "distinguishedName": null,
      "email": "exampleperson@jupiterone.com",
      "externalId": null,
      "firstname": "Example",
      "groupId": 477661,
      "id": "133771160",
      "invalidLoginAttempts": 0,
      "invitationSentAt": undefined,
      "lastLogin": undefined,
      "lastname": "Person",
      "localeCode": null,
      "lockedUntil": null,
      "managerAdId": null,
      "managerUserId": null,
      "memberOf": null,
      "name": "Example Person",
      "openidName": "example.person",
      "passwordChangedAt": undefined,
      "phone": "",
      "preferredLocaleCode": null,
      "roles": "Rock Star;Cryptozoologist;",
      "samaccountname": null,
      "state": 1,
      "status": "1",
      "title": "",
      "trustedIdpId": null,
      "updatedAt": 1623867217199,
      "username": "",
      "userprincipalname": null,
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "onelogin_user_137729105",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "activated_at": "2021-06-16T18:00:37.300Z",
            "comment": "",
            "company": "",
            "created_at": "2021-06-16T18:00:37.485Z",
            "custom_attributes": Object {
              "SuperLongAWSRoleArn": "arn:aws:iam::123456789987:role/Ninja, arn:aws:iam::123456789987:role/Developer",
              "customUserField": "AWSRole",
            },
            "department": "",
            "directory_id": null,
            "distinguished_name": null,
            "email": "exampleperson@jupiterone.com",
            "external_id": null,
            "firstname": "Example",
            "group_id": 477661,
            "id": 137729105,
            "invalid_login_attempts": 0,
            "invitation_sent_at": "2021-06-16T18:00:52.170Z",
            "last_login": "2021-08-30T21:37:36.092Z",
            "lastname": "Person",
            "locale_code": null,
            "locked_until": null,
            "manager_ad_id": null,
            "manager_user_id": null,
            "member_of": null,
            "openid_name": "example.person",
            "password_changed_at": "2021-08-30T21:37:35.405Z",
            "phone": "",
            "preferred_locale_code": null,
            "role_id": Array [
              439869,
              439870,
            ],
            "samaccountname": null,
            "state": 1,
            "status": 1,
            "title": "",
            "trusted_idp_id": null,
            "updated_at": "2021-08-30T21:37:36.196Z",
            "username": "",
            "userprincipalname": null,
          },
        },
      ],
      "_type": "onelogin_user",
      "activatedAt": 1623866437300,
      "comment": "",
      "company": "",
      "createdAt": 1623866437485,
      "customAttributes.customUserField": "AWSRole",
      "customAttributes.superLongAwsRoleArn": "arn:aws:iam::123456789987:role/Ninja, arn:aws:iam::123456789987:role/Developer",
      "department": "",
      "directoryId": null,
      "displayName": "Example Person",
      "distinguishedName": null,
      "email": "exampleperson@jupiterone.com",
      "externalId": null,
      "firstname": "Example",
      "groupId": 477661,
      "id": "137729105",
      "invalidLoginAttempts": 0,
      "invitationSentAt": 1623866452170,
      "lastLogin": 1630359456092,
      "lastname": "Person",
      "localeCode": null,
      "lockedUntil": null,
      "managerAdId": null,
      "managerUserId": null,
      "memberOf": null,
      "name": "Example Person",
      "openidName": "example.person",
      "passwordChangedAt": 1630359455405,
      "phone": "",
      "preferredLocaleCode": null,
      "roles": "Ninja;Cryptozoologist;",
      "samaccountname": null,
      "state": 1,
      "status": "1",
      "title": "",
      "trustedIdpId": null,
      "updatedAt": 1630359456196,
      "username": "",
      "userprincipalname": null,
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "onelogin_user_143505928",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "activated_at": "2021-07-29T16:15:03.946Z",
            "comment": "",
            "company": "",
            "created_at": "2021-07-29T16:15:03.954Z",
            "custom_attributes": Object {
              "SuperLongAWSRoleArn": "",
              "customUserField": "",
            },
            "department": "",
            "directory_id": null,
            "distinguished_name": null,
            "email": "exampleperson@jupiterone.com",
            "external_id": null,
            "firstname": "Example",
            "group_id": null,
            "id": 143505928,
            "invalid_login_attempts": 0,
            "invitation_sent_at": "2021-07-29T16:17:23.889Z",
            "last_login": "2021-08-30T19:19:41.139Z",
            "lastname": "Person",
            "locale_code": null,
            "locked_until": null,
            "manager_ad_id": null,
            "manager_user_id": null,
            "member_of": null,
            "openid_name": "example.person",
            "password_changed_at": "2021-08-30T19:19:40.661Z",
            "phone": "",
            "preferred_locale_code": null,
            "role_id": Array [
              459456,
            ],
            "samaccountname": null,
            "state": 1,
            "status": 1,
            "title": "",
            "trusted_idp_id": null,
            "updated_at": "2021-08-30T19:19:41.420Z",
            "username": "",
            "userprincipalname": null,
          },
        },
      ],
      "_type": "onelogin_user",
      "activatedAt": 1627575303946,
      "comment": "",
      "company": "",
      "createdAt": 1627575303954,
      "customAttributes.customUserField": "",
      "customAttributes.superLongAwsRoleArn": "",
      "department": "",
      "directoryId": null,
      "displayName": "Example Person",
      "distinguishedName": null,
      "email": "exampleperson@jupiterone.com",
      "externalId": null,
      "firstname": "Example",
      "groupId": 0,
      "id": "143505928",
      "invalidLoginAttempts": 0,
      "invitationSentAt": 1627575443889,
      "lastLogin": 1630351181139,
      "lastname": "Person",
      "localeCode": null,
      "lockedUntil": null,
      "managerAdId": null,
      "managerUserId": null,
      "memberOf": null,
      "name": "Example Person",
      "openidName": "example.person",
      "passwordChangedAt": 1630351180661,
      "phone": "",
      "preferredLocaleCode": null,
      "roles": "AWS_BINDING;",
      "samaccountname": null,
      "state": 1,
      "status": "1",
      "title": "",
      "trustedIdpId": null,
      "updatedAt": 1630351181420,
      "username": "",
      "userprincipalname": null,
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "onelogin_user_144476025",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "activated_at": "2021-08-03T18:14:05.962Z",
            "comment": "",
            "company": "",
            "created_at": "2021-08-03T18:14:05.971Z",
            "custom_attributes": Object {
              "SuperLongAWSRoleArn": "",
              "customUserField": "",
            },
            "department": "",
            "directory_id": null,
            "distinguished_name": null,
            "email": "exampleperson@jupiterone.com",
            "external_id": null,
            "firstname": "Example",
            "group_id": null,
            "id": 144476025,
            "invalid_login_attempts": 0,
            "invitation_sent_at": "2021-08-03T18:14:31.183Z",
            "last_login": "2021-08-03T18:21:58.006Z",
            "lastname": "Person",
            "locale_code": null,
            "locked_until": null,
            "manager_ad_id": null,
            "manager_user_id": null,
            "member_of": null,
            "openid_name": "example.person",
            "password_changed_at": "2021-08-03T18:21:21.695Z",
            "phone": "",
            "preferred_locale_code": null,
            "role_id": null,
            "samaccountname": null,
            "state": 1,
            "status": 1,
            "title": "",
            "trusted_idp_id": null,
            "updated_at": "2021-08-03T18:21:58.111Z",
            "username": "",
            "userprincipalname": null,
          },
        },
      ],
      "_type": "onelogin_user",
      "activatedAt": 1628014445962,
      "comment": "",
      "company": "",
      "createdAt": 1628014445971,
      "customAttributes.customUserField": "",
      "customAttributes.superLongAwsRoleArn": "",
      "department": "",
      "directoryId": null,
      "displayName": "Example Person",
      "distinguishedName": null,
      "email": "exampleperson@jupiterone.com",
      "externalId": null,
      "firstname": "Example",
      "groupId": 0,
      "id": "144476025",
      "invalidLoginAttempts": 0,
      "invitationSentAt": 1628014471183,
      "lastLogin": 1628014918006,
      "lastname": "Person",
      "localeCode": null,
      "lockedUntil": null,
      "managerAdId": null,
      "managerUserId": null,
      "memberOf": null,
      "name": "Example Person",
      "openidName": "example.person",
      "passwordChangedAt": 1628014881695,
      "phone": "",
      "preferredLocaleCode": null,
      "roles": "",
      "samaccountname": null,
      "state": 1,
      "status": "1",
      "title": "",
      "trustedIdpId": null,
      "updatedAt": 1628014918111,
      "username": "",
      "userprincipalname": null,
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "onelogin_application_1475668",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "connector_id": 68269,
            "extension": true,
            "icon": "https://s3.amazonaws.com/onelogin-assets/images/icons/doordash.png",
            "id": 1475668,
            "name": "DoorDash",
            "provisioning": false,
            "rules": Array [],
            "visible": true,
          },
        },
      ],
      "_type": "onelogin_application",
      "connectorId": 68269,
      "displayName": "DoorDash",
      "extension": true,
      "id": "1475668",
      "name": "DoorDash",
      "provisioning": false,
      "ruleIds": "",
      "visible": true,
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "onelogin_application_1531580",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "connector_id": 78887,
            "extension": false,
            "icon": "https://s3.amazonaws.com/onelogin-assets/images/icons/amazonwebservicesmultiaccount.png",
            "id": 1531580,
            "name": "Amazon Web Services (AWS) Multi Account",
            "provisioning": true,
            "rules": Array [
              Object {
                "actions": Array [
                  Object {
                    "action": "set_role",
                    "value": Array [
                      "arn:aws:iam::087679860675:role/OneLogin_EC2_ReadOnly",
                      "arn:aws:iam::087679860675:role/OneLogin_S3_ReadOnly",
                    ],
                  },
                ],
                "conditions": Array [
                  Object {
                    "operator": "ew",
                    "source": "email",
                    "value": "gmail.com",
                  },
                  Object {
                    "operator": "ri",
                    "source": "has_role",
                    "value": "459456",
                  },
                ],
                "enabled": true,
                "id": 317040,
                "match": "all",
                "name": "TestMapping",
                "position": 1,
              },
            ],
            "visible": true,
          },
        },
      ],
      "_type": "onelogin_application",
      "connectorId": 78887,
      "displayName": "Amazon Web Services (AWS) Multi Account",
      "extension": false,
      "id": "1531580",
      "name": "Amazon Web Services (AWS) Multi Account",
      "provisioning": true,
      "ruleIds": "317040",
      "visible": true,
    },
    Object {
      "_class": Array [
        "Configuration",
      ],
      "_key": "onelogin_application_rule_317040",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "actions": Array [
              Object {
                "action": "set_role",
                "value": Array [
                  "arn:aws:iam::087679860675:role/OneLogin_EC2_ReadOnly",
                  "arn:aws:iam::087679860675:role/OneLogin_S3_ReadOnly",
                ],
              },
            ],
            "conditions": Array [
              Object {
                "operator": "ew",
                "source": "email",
                "value": "gmail.com",
              },
              Object {
                "operator": "ri",
                "source": "has_role",
                "value": "459456",
              },
            ],
            "enabled": true,
            "id": 317040,
            "match": "all",
            "name": "TestMapping",
            "position": 1,
          },
        },
      ],
      "_type": "onelogin_application_rule",
      "actions": "[{\\"action\\":\\"set_role\\",\\"value\\":[\\"arn:aws:iam::087679860675:role/OneLogin_EC2_ReadOnly\\",\\"arn:aws:iam::087679860675:role/OneLogin_S3_ReadOnly\\"]}]",
      "conditions": "[{\\"source\\":\\"email\\",\\"operator\\":\\"ew\\",\\"value\\":\\"gmail.com\\"},{\\"source\\":\\"has_role\\",\\"operator\\":\\"ri\\",\\"value\\":\\"459456\\"}]",
      "displayName": "TestMapping",
      "enabled": true,
      "id": "317040",
      "match": "all",
      "name": "TestMapping",
      "position": 1,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin:sso:Local Integration",
      "_toEntityKey": "onelogin:sso:Local Integration",
      "_type": "onelogin_account_has_service",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin:mfa:Local Integration",
      "_toEntityKey": "onelogin:mfa:Local Integration",
      "_type": "onelogin_account_has_service",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_group_477661",
      "_toEntityKey": "onelogin_group_477661",
      "_type": "onelogin_account_has_group",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_428308",
      "_toEntityKey": "onelogin_role_428308",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_439868",
      "_toEntityKey": "onelogin_role_439868",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_439869",
      "_toEntityKey": "onelogin_role_439869",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_439870",
      "_toEntityKey": "onelogin_role_439870",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_459456",
      "_toEntityKey": "onelogin_role_459456",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_user_132422371",
      "_toEntityKey": "onelogin_user_132422371",
      "_type": "onelogin_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_132422371",
      "_key": "onelogin_user_132422371|assigned|onelogin_role_439868",
      "_toEntityKey": "onelogin_role_439868",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_user_133771160",
      "_toEntityKey": "onelogin_user_133771160",
      "_type": "onelogin_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_group_477661",
      "_key": "onelogin_group_477661|has|onelogin_user_133771160",
      "_toEntityKey": "onelogin_user_133771160",
      "_type": "onelogin_group_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_133771160",
      "_key": "onelogin_user_133771160|assigned|onelogin_group_477661",
      "_toEntityKey": "onelogin_group_477661",
      "_type": "onelogin_user_assigned_group",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_133771160",
      "_key": "onelogin_user_133771160|assigned|onelogin_role_439868",
      "_toEntityKey": "onelogin_role_439868",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_133771160",
      "_key": "onelogin_user_133771160|assigned|onelogin_role_439870",
      "_toEntityKey": "onelogin_role_439870",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_user_137729105",
      "_toEntityKey": "onelogin_user_137729105",
      "_type": "onelogin_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_group_477661",
      "_key": "onelogin_group_477661|has|onelogin_user_137729105",
      "_toEntityKey": "onelogin_user_137729105",
      "_type": "onelogin_group_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_137729105",
      "_key": "onelogin_user_137729105|assigned|onelogin_group_477661",
      "_toEntityKey": "onelogin_group_477661",
      "_type": "onelogin_user_assigned_group",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_137729105",
      "_key": "onelogin_user_137729105|assigned|onelogin_role_439869",
      "_toEntityKey": "onelogin_role_439869",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_137729105",
      "_key": "onelogin_user_137729105|assigned|onelogin_role_439870",
      "_toEntityKey": "onelogin_role_439870",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_user_143505928",
      "_toEntityKey": "onelogin_user_143505928",
      "_type": "onelogin_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_143505928",
      "_key": "onelogin_user_143505928|assigned|onelogin_role_459456",
      "_toEntityKey": "onelogin_role_459456",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_user_144476025",
      "_toEntityKey": "onelogin_user_144476025",
      "_type": "onelogin_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_application_1475668",
      "_toEntityKey": "onelogin_application_1475668",
      "_type": "onelogin_account_has_application",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_application_1531580",
      "_toEntityKey": "onelogin_application_1531580",
      "_type": "onelogin_account_has_application",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_application_1531580",
      "_key": "onelogin_application_1531580|has|onelogin_application_rule_317040",
      "_toEntityKey": "onelogin_application_rule_317040",
      "_type": "onelogin_application_has_rule",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_133771160",
      "_key": "onelogin_user_133771160|assigned|onelogin_application_1475668",
      "_toEntityKey": "onelogin_application_1475668",
      "_type": "onelogin_user_assigned_application",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_137729105",
      "_key": "onelogin_user_137729105|assigned|onelogin_application_1475668",
      "_toEntityKey": "onelogin_application_1475668",
      "_type": "onelogin_user_assigned_application",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_143505928",
      "_key": "onelogin_user_143505928|assigned|onelogin_application_1531580",
      "_toEntityKey": "onelogin_application_1531580",
      "_type": "onelogin_user_assigned_application",
      "displayName": "ASSIGNED",
    },
  ],
  "encounteredTypes": Array [
    "onelogin_account",
    "onelogin_service",
    "onelogin_account_has_service",
    "onelogin_group",
    "onelogin_account_has_group",
    "onelogin_role",
    "onelogin_account_has_role",
    "onelogin_user",
    "onelogin_account_has_user",
    "onelogin_user_assigned_role",
    "onelogin_group_has_user",
    "onelogin_user_assigned_group",
    "onelogin_application",
    "onelogin_account_has_application",
    "onelogin_application_rule",
    "onelogin_application_has_rule",
    "onelogin_user_assigned_application",
  ],
  "numCollectedEntities": 17,
  "numCollectedRelationships": 29,
}
`;
